### PR TITLE
Refactor, implement OpStore interface

### PR DESCRIPTION
**Pull request by DavidBuchanan314**
Previously, the `LogValidationContext` struct stored all the state required for verifying operation logs. It was also responsible for some aspects of operation validation (e.g. 72h limit, nullification checks, etc.) but *not* others (e.g. signature verification).

NOTE: the API description here is stale, see review comments / the code for up-to-date details.

This PR introduces the `OpStore` interface, which abstracts the storage layer, along with an `InMemoryOpStore` implementation. (Separately, I have postgres and sqlite implementations of the interface via gorm, as part of the replica service impl)

```go
type OpStore interface {
	// GetHead returns the CID of the most recent valid operation for a DID.
	// Returns empty string if the DID does not exist.
	GetHead(ctx context.Context, did string) (string, error)

	// GetMetadata returns metadata about a specific operation.
	// Returns an error if the operation does not exist.
	GetMetadata(ctx context.Context, did string, cid string) (*OpStatus, error)

	// GetOperation returns the operation data for a specific DID and CID.
	// Returns an error if the operation does not exist.
	GetOperation(ctx context.Context, did string, cid string) (Operation, error)

	// CommitOperations atomically commits a batch of prepared operations to the store.
	// All operations in the batch are committed or none are (all-or-nothing).

	// For each PreparedOperation, `prevHead` MUST match the head value returned by an earlier call to GetHead.
	// If multiple updates to the same DID are attempted concurrently, one will return an error due to head mismatch.
	CommitOperations(ctx context.Context, ops []*PreparedOperation) error
}
```

```go 
func VerifyOperation(ctx context.Context, store OpStore, did string, op Operation, createdAt time.Time) (*PreparedOperation, error) {
```

The `VerifyOperation` method uses an `OpStore` to build a verified `PreparedOperation` struct, which represents a description of changes to be later applied to the store via a call to `CommitOperations`. `CommitOperations` takes a slice of `PreparedOperation`s so that commits can be batched (optionally), which supports bulk verification use cases (e.g. replica backfill).

`VerifyOpLog` has been updated to use `OpStore` instead of `LogValidationContext`, which shows a basic example of how it's used (which is much simpler!). Aside from syntactic validation, almost all the verification logic now happens inside `VerifyOperation`

### CODE REVIEW

File: operation.go [hunk 1]
```diff
@@ -528,3 +528,25 @@ func (oe *OpEnum) AsOperation() Operation {
...
+	case *LegacyOp:
+		opEnum.Legacy = v
+	case *TombstoneOp:
+		opEnum.Tombstone = v
+	default:
+		return nil, fmt.Errorf("unknown operation type: %T", op)
+	}
+
+	return opEnum, nil
+}
```
Comment (DavidBuchanan314) [inline operation.go]: This change is unrelated to the OpStore interface, it's just something else I'm using in the replica impl

File: opstore.go [hunk 2]
```diff
@@ -0,0 +1,317 @@
...
+
+	// CommitOperations atomically commits a batch of prepared operations to the store.
+	// All operations in the batch are committed or none are (all-or-nothing).
+
+	// For each PreparedOperation, `prevHead` MUST match the head value returned by an earlier call to GetHead.
+	// If multiple updates to the same DID are attempted concurrently, one will return an error due to head mismatch.
+	CommitOperations(ctx context.Context, ops []*PreparedOperation) error
+}
+
+type InMemoryOpStore struct {
```
Comment (bnewbold) [inline opstore.go]: As a small style opinion, I think `MemOpStore` is a bit more idiomatic for an implementation that persists in RAM.

Not a huge deal either way.

File: operation.go [hunk 3]
```diff
@@ -484,6 +485,31 @@ func (o *OpEnum) UnmarshalJSON(b []byte) error {
 	}
 }
 
+// Value implements the driver.Valuer interface
```
Comment (bnewbold) [inline operation.go]: I think i'd recommend against implementing serialization like this. We already have JSON marshaling working, so implementing serialization shouldn't be too hard in `OpStore` implementations.

The reason is that I don't think the `database/sql/driver` interface has all that much adoption these days, and having generic `Value()` and `Scan()` implementations could cause confusion or even remove flexibility.

I'd defer if other folks think it is helpful though.

File: operation.go [hunk 4]
```diff
@@ -528,3 +554,25 @@ func (oe *OpEnum) AsOperation() Operation {
 		return nil
 	}
 }
+
+// WrapOperation converts an Operation interface to an OpEnum for marshaling
+func WrapOperation(op Operation) (*OpEnum, error) {
```
Comment (bnewbold) [inline operation.go]: This could be a method on the `Operation` interface instead? Or name it `NewOpEnum(op Operation)`? I wouldn't know/remember what `WrapOperation(op)` does without looking up the function signature.

Not a big thing or blocker.

File: opstore.go [hunk 5]
```diff
@@ -0,0 +1,317 @@
...
+	DID          string
+	PrevHead     string
+	NullifiedOps []string // CIDs of any operations being nullified
+	KeyIndex     int
+	CreatedAt    time.Time
+	Op           Operation
+	OpCid        string
+}
+
+type OpStore interface {
```
Comment (bnewbold) [inline opstore.go]: I might update this comment after reviewing the replica code, but my initial feeling is that this interface could use a bit of iteration.

It feels like `OpStatus` could be replaced with something like an `OpEntry` which contains both the metadata and the raw operation itself, similar to the output of the `/{did}/log/audit` endpoint. It should include the `OpCID`.

Instead of needing both `GetMetadata` and `GetOperation`, could have a single `GetEntry(ctx, did, cid) (*OpEntry, error)` which returns both together.

Instead of `GetHead`, do `GetLatest(ctx, did) (*OpEntry, error)`. That avoids additional `GetMetadata` and/or `GetOperation` calls; and if those weren't necessary, the overhead should not be much.

I think `CommitOperations` is generally good: takes a batch, does everything as a transaction. It feels like `PreparedOperation` could be an `OpEntry` plus one or two fields?

Feels like the interface should include `GetAllEntries(ctx, did) ([]OpEntry, error)` which includes all operations and metdata, even if nullified. I guess maybe we'd maybe want to paginate that at some point like `GetEntriesSince(ctx, did, index)` or something.

I could be wrong, but the above is my intuition of what will work well for SQL stores, and avoid race-y corner-cases like `GetHead()`, but then by the time you call `GetMetadata` with that CID the row has already been nullified.

**Review by bnewbold — COMMENTED** [review]
I like the general approach here.

Left some comments on the interface/API shape; hoping it can be tweaked to be a tighter without too much refactoring trouble.

I haven't reviewed the implementation in detail yet. A couple more tests around corner-cases would probably be good, even just against the in-memory implementation. Eg, conflicting calls to `CommitOperations()` for the same DID.
